### DISCUSSION TITLE
Disable scrolling to the top after navigation

### DISCUSSION THREAD

**Discussion by N-B-H** [swup/swup#754] [Help] [answered]
Hi,
I am trying to use swup with astro. I tried both, the new astro integration (which I am happy to see!) and standard installation with Preload and Scroll Plugins.

I want a basic thing: To scroll to a certain position after page transitioning, so that the new content is nicely in view.

Navigation seems to scroll up to the top of each page, no matter what options I set. 
This wasn't the case in some earlier version of swup, but going back to swup 3 doesn't resolve my issue. Maybe it was a very specific release which I don't remember now.

Anyway, I tried setting the offset:

```js
{
  offset: () => document.querySelector('#heroimg').offsetHeight * 0.85,
}
```
quite like in the documentation, but this didn't change anything, maybe i understood it's functionality wrong.

Same for 
```js
{
  shouldResetScrollPosition: (link) => false, 
}
```
I tried scrolling to my intended position by hand with hooks, as commented at the bottom of my code below, but this would require maintaining the current position and scroll from there. I can't turn off going up to the top and my own scrolling still only happens after swup went up to the top of the page (with or without scroll plugin) and not from the previous position to the new desired position. Using other lifecycle hooks didn't resolve the issue. Swup seems to really like the top of my page.

I must obviously be missing something, since my use case feels quite basic, but I can't find it, unfortunately.

any ideas?

Thanks and all best,

Niklas

my code with the offset, which i would think should be working:

```html
      <main
        id="swup"
        class="transition-fade mx-auto flex max-w-3xl flex-col gap-4 p-10 md:max-w-4xl" 
      >

        <slot />
      </main>
      <script>
        import Swup from "swup";
        import SwupPreloadPlugin from "@swup/preload-plugin";
        import SwupScrollPlugin from "@swup/scroll-plugin";

        const swup = new Swup({
          plugins: [
            new SwupPreloadPlugin(),
            new SwupScrollPlugin({
              // doScrollingRightAway: true,
              // animateScroll: {
              //   betweenPages: false, 
              //   samePageWithHash: false,
              //   samePage: false,
              // },
              offset: () => document.querySelector('#heroimg').offsetHeight * 0.85, // <---- does nothing, still goes up to the top
              //   shouldResetScrollPosition: (link) => false, // <---- still resets scroll position to the top
            }),
          ],
        });

        // document.addEventListener("swup:visit:end", () => {
        //   // window.scroll({top: document.querySelector("#heroimg").offsetHeight*0.85, behavior: "smooth"});
        //   swup.scrollTo(document.querySelector("#heroimg").offsetHeight * 0.85);
        // });
        // document.addEventListener("swup:link:self", () => {
        //   // window.scroll({top: document.querySelector("#heroimg").offsetHeight*0.85, behavior: "smooth"});
        //   swup.scrollTo(document.querySelector("#heroimg").offsetHeight * 0.85);
        // });
      </script>
```

**Comment by hirasso** [comment]
@N-B-H as a first test, you could target the element you want to scroll to from your links, for example:

```html
<a href="/my-page/#swup">My Page</a> 
```
Swup should then automatically scroll to the targeted element as soon as the page is rendered, just like the browser would.

If you don't want to have the hash in the URLs, it would be helpful to work with a reduced test case, so that we know exactly what we are talking about. You could fork any of the [demos](https://swup.js.org/getting-started/demos/) on Replit and go from there.

**Comment by daun** [comment, marked as answer]
You can turn off the scroll reset inside a hook, in which case swup will simply do nothing and you're free to implement that behavior yourself. See the example in [Disable scroll reset](https://swup.js.org/visit/#disable-scroll-reset). 

```js
swup.hooks.on('visit:start', (visit) => {
  visit.scroll.reset = false;
});
```

I think the best thing in your specific case would be manually setting the scroll target to your actual content on each visit. That way, swup will treat the new page as if a link to that specific anchor was clicked. This will override any actual anchors in clicked links, so you might want to wrap it in an `if` statement.

```js
swup.hooks.on('visit:start', (visit) => {
  visit.scroll.target = '#main-content';
});
```

The built-in scroll support is new in swup 4, so I'd assume that reverting to swup 3 should disable any scrolling -- assuming you're not adding the scroll plugin later anyway.

**Comment by hirasso** [comment]
I went ahead and created a test case for the second option suggested by @daun :

https://replit.com/@swupjs/Swup-Test-Scrolling#swup.js

**Comment by N-B-H** [comment]
Wow. Thank you so much!

I overlooked the possibility of visit.scroll.reset = false; in the docs before, my bad. This is exactly what I was looking for.
Since I have a fixed overlay/header at the top, exactly scrolling to the element as an anchor would scroll it just underneath the overlay, so I'd still have to adjust an offset. But, disabling the scroll reset allows me to go exactly to the calculated Y-Location that I want, and when I want. The page:view and link:self -hooks seem to work just fine.

This works as planned:

```html
<script>
  // Initialize Swup with plugins
  import Swup from "swup";
  import SwupPreloadPlugin from "@swup/preload-plugin";
  import SwupScrollPlugin from "@swup/scroll-plugin";

  const swup = new Swup({
    plugins: [
      new SwupPreloadPlugin(),
      new SwupScrollPlugin(),
    ],
  });
  
  // calculate the scroll position, based on hero image and header, because its height might change
  const header = document.querySelector("header");
  const heroimg = document.querySelector("#heroimg");
  
  function getScrollPosition() {
    return heroimg.offsetHeight - header.offsetHeight + 1 // chrome needs a magic +1 in my case...
  };

  // disable the default scrolling to the top
  swup.hooks.on("visit:start", (visit) => {
    visit.scroll.reset = false;
  });

  // scroll to calculated position on both: 
  // when navigating to other pages and when clicking the current link again
  swup.hooks.on("page:view", (visit) => {
    swup.scrollTo(getScrollPosition());
  });
  
  swup.hooks.on("link:self", (visit) => {
    swup.scrollTo(getScrollPosition());
  });        
</script>
```


Thanks again for the quick help! 
I am rather new to this and after spending some hours in struggle and frustration yesterday, You guys now really made my day! :)
It's a great library. Thank you for making and maintaining it.

All the best wishes,
Niklas

**Reply by daun** [reply]
Great to know you have it working!
